View transitions don't capture composited filters in the old snapshot.

mattwoodrow opened the pull request:
#### a73751a1a8cbe720473262dc7b04fb874aff3e8e
<pre>
View transitions don&apos;t capture composited filters in the old snapshot.
<a href="https://bugs.webkit.org/show_bug.cgi?id=293393">https://bugs.webkit.org/show_bug.cgi?id=293393</a>
&lt;<a href="https://rdar.apple.com/151771464">rdar://151771464</a>&gt;

Reviewed by Said Abou-Hallawa.

RenderLayer::paintsWithFilter needs to take
PaintBehavior::FlattenCompositingLayers into account for the filters to be
painted into the snapshot.

RenderLayer::m_filters is retained state that depends on paintsWithFilter, so
make sure we initialize it when needed for snapshot painting (and clear again
afterwards if necessary).

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/inline-child-with-composited-filter-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/inline-child-with-composited-filter-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/inline-child-with-composited-filter.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintsWithFilters const):
(WebCore::RenderLayer::hasFiltersForPainting const):
(WebCore::RenderLayer::filtersForPainting):
(WebCore::RenderLayer::setupFilters):
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::calculateClipRects const):
(WebCore::RenderLayer::filtersForPainting const): Deleted.
* Source/WebCore/rendering/RenderLayer.h:

Canonical link: <a href="https://commits.webkit.org/295357@main">https://commits.webkit.org/295357@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79280ecb44eb7d4ce0920cb448e65d974adee71f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104774 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24486 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14908 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109988 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55448 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106814 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24877 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33032 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79555 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107780 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19340 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94555 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59862 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19099 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12632 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54830 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88796 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12679 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112390 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31939 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23485 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88634 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32303 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90781 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88258 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22510 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33150 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10919 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27252 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31864 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37218 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31656 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34997 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33215 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3569,17 +3572,27 @@ void RenderLayer::clearLayerClipPath()
+    if (!hasFiltersForPainting(context, paintFlags, paintBehavior))
         return nullptr;
 
+    updateFilterPaintingStrategy(paintBehavior);
```
`updateFilterPaintingStrategy()` will call `paintsWithFilters(paintBehavior)` which will return true because `hasFiltersForPainting ()` already returns true. Can't we refactor `updateFilterPaintingStrategy()` such that we do not call `paintsWithFilters(paintBehavior)` twice?

shallawa replied:
I would suggest moving these two lines to `ensureLayerFilters()`
```
    m_filters->setPreferredFilterRenderingModes(renderer().page().preferredFilterRenderingModes());
    m_filters->setFilterScale({ page().deviceScaleFactor(), page().deviceScaleFactor() });
```
So this function can call `ensureLayerFilters()` instead of calling `updateFilterPaintingStrategy()`.

shallawa commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3569,17 +3572,27 @@ void RenderLayer::clearLayerClipPath()
 }
 
-RenderLayerFilters* RenderLayer::filtersForPainting(GraphicsContext& context, OptionSet<PaintLayerFlag> paintFlags) const
+bool RenderLayer::hasFiltersForPainting(GraphicsContext& context, OptionSet<PaintLayerFlag> paintFlags, OptionSet<PaintBehavior> paintBehavior) const
```
Now we have two functions answering almost the same question and the difference between them is very subtle: `hasFiltersForPainting()` and `paintsWithFilters()`. Can we rename them such that the names reflect exactly what they return?

shallawa replied:
How about these names?
```
RenderLayer::shouldHaveFilters(OptionSet<PaintBehavior> paintBehavior)
RenderLayer::shouldHaveFiltersForPainting(GraphicsContext& context, OptionSet<PaintLayerFlag> paintFlags, OptionSet<PaintBehavior> paintBehavior)
```

mattwoodrow replied:
I think keeping the 'painting' bit of the name is important, since is distinct from having filters but using them as a property on a compositing layer.

shallawa commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1022,6 +1022,9 @@ bool RenderLayer::paintsWithFilters() const
+    if (paintBehavior & PaintBehavior::FlattenCompositingLayers)
+        return true;
+
     return !m_backing->canCompositeFilters();
```
The original code was a little bit hard to understand and this new condition makes it even harder. Does `paintBehavior` have to have `PaintBehavior::FlattenCompositingLayers` in addition to `isComposited()` true? Before your change I think the following code should be clearer:

```
if (isComposited() && m_backing->canCompositeFilters())
	return false;

return true;
```
Can your new condition fit with change?

shallawa commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5639,7 +5656,7 @@ LayoutRect RenderLayer::calculateLayerBounds(const RenderLayer* ancestorLayer, c
         computeLayersUnion(*childLayer);
 
-    if (flags.contains(IncludeFilterOutsets) || (flags.contains(IncludePaintedFilterOutsets) && paintsWithFilters()))
+    if (flags.contains(IncludeFilterOutsets) || (flags.contains(IncludePaintedFilterOutsets) && (paintsWithFilters() || (flags & IncludeCompositedDescendants))))
```
```
... paintsWithFilters(flags & IncludeCompositedDescendants ? PaintBehavior::FlattenCompositingLayers : OptionSet<PaintBehavior>()))
```

shallawa commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3569,17 +3572,27 @@ void RenderLayer::clearLayerClipPath()
+
     if (m_filters)
         return m_filters.get();
 
```
Not new but 

```
    if (m_filters)
        return m_filters.get();
    return nullptr;
````
can be replaced by
```
    return m_filters.get();
```

shallawa replied:
Another suggestion:

```
RenderLayerFilters& RenderLayer::ensureLayerFilters()
{
    if (m_filters)
        return *m_filters;
    
    m_filters = makeUnique<RenderLayerFilters>(*this);
    m_filters->setPreferredFilterRenderingModes(renderer().page().preferredFilterRenderingModes());
    m_filters->setFilterScale({ page().deviceScaleFactor(), page().deviceScaleFactor() });
    return *m_filters;
}

RenderLayerFilters* RenderLayer::filtersForPainting(GraphicsContext& context, OptionSet<PaintLayerFlag> paintFlags, OptionSet<PaintBehavior> paintBehavior)
{
    if (!shouldHaveFiltersForPainting(context, paintFlags, paintBehavior))
        return nullptr;

    return &ensureLayerFilters();
}
````

shallawa commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3569,26 +3572,31 @@ void RenderLayer::clearLayerClipPath()
-    return nullptr;
+RenderLayerFilters* RenderLayer::filtersForPainting(GraphicsContext& context, OptionSet<PaintLayerFlag> paintFlags, OptionSet<PaintBehavior> paintBehavior)
+{
+    if (!hasFiltersForPainting(context, paintFlags, paintBehavior))
```
What is the question here:
1. Does `RenderLayer` have `RenderLayerFilters` for painting? If yes then `hasFiltersForPainting()` is fine name
2. Should `RenderLayer` have `RenderLayerFilters` for painting? If yes then `hasFiltersForPainting()` is not a matching name

Given that we are calling `ensureLayerFilters()` below, I think `hasFiltersForPainting()` is not a matching name.

mattwoodrow replied:
That's fair, changing it to `shouldHaveFiltersForPainting`.

shallawa commented on LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/inline-child-with-composited-filter.html:
```diff
@@ -0,0 +1,56 @@
+<html class=reftest-wait>
+<title>View transitions: inline child with filter</title>
+<link rel="help" href="https://www.w3.org/TR/css-view-transitions-1/">
+<link rel="author" href="mailto:vmpstr@chromium.org">
```
Is this a new test or we just importing it? Do we need to add it to w3c-import.log?

mattwoodrow replied:
It's a new test, but I left the original author there since it's mostly a duplicate of an existing test (with the addition of forcing compositing the filtered element).

mattwoodrow committed:
View transitions don't capture composited filters in the old snapshot.
https://bugs.webkit.org/show_bug.cgi?id=293393
<rdar://151771464>

Reviewed by Said Abou-Hallawa.

RenderLayer::paintsWithFilter needs to take
PaintBehavior::FlattenCompositingLayers into account for the filters to be
painted into the snapshot.

RenderLayer::m_filters is retained state that depends on paintsWithFilter, so
make sure we initialize it when needed for snapshot painting (and clear again
afterwards if necessary).

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/inline-child-with-composited-filter-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/inline-child-with-composited-filter-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/inline-child-with-composited-filter.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintsWithFilters const):
(WebCore::RenderLayer::hasFiltersForPainting const):
(WebCore::RenderLayer::filtersForPainting):
(WebCore::RenderLayer::setupFilters):
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::calculateClipRects const):
(WebCore::RenderLayer::filtersForPainting const): Deleted.
* Source/WebCore/rendering/RenderLayer.h:

Canonical link: https://commits.webkit.org/295357@main
